addons/binding/org.openhab.binding.denonmarantz/src/main/java/org/openhab/binding/denonmarantz/handler/DenonMarantzHandler.java

```diff
@@ -0,0 +1,208 @@
+                    break;
+                case CHANNEL_MAIN_VOLUME:
+                    connector.sendVolumeCommand(command, 1);
+                    break;
+                case CHANNEL_MAIN_VOLUME_DB:
+                    connector.sendVolumeDbCommand(command, 1);
+                    break;
+                case CHANNEL_INPUT:
+                    connector.sendInputCommand(command, 1);
+                    break;
+                case CHANNEL_COMMAND:
+                    connector.sendCustomCommand(command);
+                    break;
+
+                case CHANNEL_ZONE2_POWER:
+                    connector.sendPowerCommand(command, 2);
+                    break;
+                case CHANNEL_ZONE2_MUTE:
+                    connector.sendMuteCommand(command, 2);
+                    break;
+                case CHANNEL_ZONE2_VOLUME:
+                    connector.sendVolumeCommand(command, 2);
+                    break;
+                case CHANNEL_ZONE2_VOLUME_DB:
+                    connector.sendVolumeCommand(command, 2);
+                    break;
+                case CHANNEL_ZONE2_INPUT:
+                    connector.sendInputCommand(command, 2);
+                    break;
+
+                case CHANNEL_ZONE3_POWER:
+                    connector.sendPowerCommand(command, 3);
+                    break;
+                case CHANNEL_ZONE3_MUTE:
+                    connector.sendMuteCommand(command, 3);
+                    break;
+                case CHANNEL_ZONE3_VOLUME:
+                    connector.sendVolumeCommand(command, 3);
+                    break;
+                case CHANNEL_ZONE3_VOLUME_DB:
+                    connector.sendVolumeCommand(command, 3);
+                    break;
+                case CHANNEL_ZONE3_INPUT:
+                    connector.sendInputCommand(command, 3);
+                    break;
+
+                default:
+                    logger.warn("Command for channel {} not supported.", channelUID.getId());
+                    break;
+            }
+
+        } catch (UnsupportedCommandTypeException e) {
+            logger.warn("Unsupported command {} for channel {}", command, channelUID.getId());
+        }
+    }
+
+    @Override
+    public void initialize() {
+        // create connection (either Telnet & HTTP or HTTP only)
+        // ThingStatus ONLINE/OFFLINE is set when AVR status is known.
+        createConnection();
+        addZoneChannels();
+    }
+
+    private void createConnection() {
+        configuration = getConfigAs(DenonMarantzConfiguration.class);
+        if (connector != null) {
+            connector.dispose();
+        }
+        connector = new DenonMarantzConnector(configuration, this, scheduler);
+        connector.connect();
+    }
+
+    private void addZoneChannels() {
+        // only add when this is a new Thing instance (and doesn't have the 'zones' property set yet).
+
+        if (this.getThing().getProperties().containsKey("zones")) {
```

Comment:
I'm not too sure this is a right approach..
My intention was:
Some Denon/Marantz models have a 2nd zone, and some have an additional 3rd zone. I only want to add the specific channels when the AVR has these zones.
In the Thing property 'zones' (not to be confused with config parameter) I store the number of configured zones after adding the channels. So I use the property as a 'guard' to detect whether the channels were already added before.

With all knowledge gained in the meantime, I think it's better to detect the number of zones during Discovery and, in case of manual adding the Thing, to let the user override the default zones parameter (2) when needed.

With the Telnet connection there is no (convenient) way to detect the number of zones.

Will try to rewrite this part.

Reply:
Having the property as a guard actually seems like a HACK to me.

So yes, discovering the channels within the discovery sounds plausible.

For the Telnet way: Why can't you discover the zones with it? What is the problem there?

Reply:
With Telnet we can only watch/query/set state, we cannot query properties. In theory could try to query the `Z2` and `Z3` state. When it replies within a short timeframe we could detect it has those zones.

Reply:
Na this sounds like a HACK to me too, better leave out zone detection via telnet then.